File: idg_backend/placed_students/certificate_generator.py
"""
Certificate Generator for Placed Students
This script generates PDF certificates for placed students
"""

from reportlab.lib.pagesizes import A4, landscape
from reportlab.lib.units import inch
from reportlab.pdfgen import canvas
from reportlab.lib import colors
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.platypus import Paragraph
from reportlab.lib.enums import TA_CENTER
from io import BytesIO
from datetime import datetime


def generate_certificate(student_name, student_id, certificate_id, company_name, role):
    """
    Generate a certificate PDF for a placed student
    
    Args:
        student_name: Name of the student
        student_id: Unique student ID (e.g., IDS_001)
        certificate_id: Unique certificate ID (e.g., IDSC_001)
        company_name: Name of the company where student is placed
        role: Role/designation of the student
    
    Returns:
        BytesIO object containing the PDF
    """
    buffer = BytesIO()
    
    # Create PDF with landscape A4 size
    page_width, page_height = landscape(A4)
    c = canvas.Canvas(buffer, pagesize=landscape(A4))
    
    # Draw border
    c.setStrokeColor(colors.HexColor('#667eea'))
    c.setLineWidth(3)
    c.rect(30, 30, page_width - 60, page_height - 60, stroke=1, fill=0)
    
    # Inner border
    c.setStrokeColor(colors.HexColor('#f7f7f7'))
    c.setLineWidth(10)
    c.rect(40, 40, page_width - 80, page_height - 80, stroke=1, fill=0)
    
    # Draw decorative corners
    corner_size = 50
    c.setStrokeColor(colors.HexColor('#667eea'))
    c.setLineWidth(2)
    
    # Top-left corner
    c.line(60, page_height - 60, 60 + corner_size, page_height - 60)
    c.line(60, page_height - 60, 60, page_height - 60 - corner_size)
    
    # Top-right corner
    c.line(page_width - 60, page_height - 60, page_width - 60 - corner_size, page_height - 60)
    c.line(page_width - 60, page_height - 60, page_width - 60, page_height - 60 - corner_size)
    
    # Bottom-left corner
    c.line(60, 60, 60 + corner_size, 60)
    c.line(60, 60, 60, 60 + corner_size)
    
    # Bottom-right corner
    c.line(page_width - 60, 60, page_width - 60 - corner_size, 60)
    c.line(page_width - 60, 60, page_width - 60, 60 + corner_size)
    
    # Logo/Header
    c.setFont("Helvetica-Bold", 36)
    c.setFillColor(colors.HexColor('#667eea'))
    c.drawCentredString(page_width / 2, page_height - 100, "iDigitalStudies")
    
    # Certificate Title
    c.setFont("Helvetica-Bold", 42)
    c.setFillColor(colors.HexColor('#333333'))
    c.drawCentredString(page_width / 2, page_height - 150, "CERTIFICATE OF ACHIEVEMENT")
    
    # Subtitle
    c.setFont("Helvetica", 18)
    c.setFillColor(colors.HexColor('#666666'))
    c.drawCentredString(page_width / 2, page_height - 185, "This is to certify that")
    
    # Student Name
    c.setFont("Helvetica-Bold", 36)
    c.setFillColor(colors.HexColor('#667eea'))
    c.drawCentredString(page_width / 2, page_height - 240, student_name)
    
    # Underline for name
    name_width = c.stringWidth(student_name, "Helvetica-Bold", 36)
    c.setStrokeColor(colors.HexColor('#667eea'))
    c.setLineWidth(2)
    c.line(page_width / 2 - name_width / 2, page_height - 245, 
           page_width / 2 + name_width / 2, page_height - 245)
    
    # Body Text
    c.setFont("Helvetica", 16)
    c.setFillColor(colors.HexColor('#555555'))
    
    y_position = page_height - 285
    c.drawCentredString(page_width / 2, y_position, "has successfully completed the Digital Marketing program")
    
    y_position -= 25
    c.drawCentredString(page_width / 2, y_position, "and has been placed at")
    
    # Company Name
    y_position -= 35
    c.setFont("Helvetica-Bold", 20)
    c.setFillColor(colors.HexColor('#667eea'))
    c.drawCentredString(page_width / 2, y_position, company_name)
    
    # Role
    y_position -= 30
    c.setFont("Helvetica", 16)
    c.setFillColor(colors.HexColor('#555555'))
    c.drawCentredString(page_width / 2, y_position, "as a")
    
    y_position -= 30
    c.setFont("Helvetica-Bold", 20)
    c.setFillColor(colors.HexColor('#667eea'))
    c.drawCentredString(page_width / 2, y_position, role)
    
    # Congratulations text
    y_position -= 45
    c.setFont("Helvetica", 14)
    c.setFillColor(colors.HexColor('#555555'))
    first_name = student_name.split()[0]
    c.drawCentredString(page_width / 2, y_position, 
                       f"We congratulate {first_name} on this remarkable achievement")
    y_position -= 20
    c.drawCentredString(page_width / 2, y_position, 
                       "and wish them continued success in their professional journey.")
    
    # Footer with signatures
    footer_y = 120
    
    # Left signature
    left_x = page_width / 4
    c.setStrokeColor(colors.HexColor('#333333'))
    c.setLineWidth(1)
    c.line(left_x - 80, footer_y, left_x + 80, footer_y)
    c.setFont("Helvetica-Bold", 12)
    c.setFillColor(colors.HexColor('#333333'))
    c.drawCentredString(left_x, footer_y - 20, "Director")
    c.setFont("Helvetica", 10)
    c.setFillColor(colors.HexColor('#666666'))
    c.drawCentredString(left_x, footer_y - 35, "iDigitalStudies")
    
    # Center - Date and Student ID
    center_x = page_width / 2
    c.setFont("Helvetica-Bold", 11)
    c.setFillColor(colors.HexColor('#333333'))
    current_date = datetime.now().strftime("%B %d, %Y")
    c.drawCentredString(center_x, footer_y + 10, f"Date: {current_date}")
    c.drawCentredString(center_x, footer_y - 10, f"Student ID: {student_id}")
    
    # Right signature
    right_x = 3 * page_width / 4
    c.setStrokeColor(colors.HexColor('#333333'))
    c.setLineWidth(1)
    c.line(right_x - 80, footer_y, right_x + 80, footer_y)
    c.setFont("Helvetica-Bold", 12)
    c.setFillColor(colors.HexColor('#333333'))
    c.drawCentredString(right_x, footer_y - 20, "Head of Training")
    c.setFont("Helvetica", 10)
    c.setFillColor(colors.HexColor('#666666'))
    c.drawCentredString(right_x, footer_y - 35, "iDigitalStudies")
    
    # Certificate ID at bottom right
    c.setFont("Courier", 9)
    c.setFillColor(colors.HexColor('#999999'))
    c.drawRightString(page_width - 60, 50, f"Certificate ID: {certificate_id}")
    
    # Finalize PDF
    c.showPage()
    c.save()
    
    buffer.seek(0)
    return buffer


def save_certificate_to_file(student_name, student_id, certificate_id, company_name, role, filename):
    """
    Generate and save certificate to a file
    """
    buffer = generate_certificate(student_name, student_id, certificate_id, company_name, role)
    
    with open(filename, 'wb') as f:
        f.write(buffer.getvalue())
    
    return filename

File: nextjs-app/lib/certificateGenerator.ts
/**
 * Certificate Generator Utility
 * This utility generates a certificate image/PDF for placed students
 */

export interface CertificateData {
  studentName: string;
  studentId: string;
  certificateId: string;
  companyName: string;
  role: string;
  date?: string;
}

export const generateCertificateHTML = (data: CertificateData): string => {
  const currentDate = data.date || new Date().toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'long',
    day: 'numeric'
  });

  return `
<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <style>
    @page {
      size: A4 landscape;
      margin: 0;
    }
    
    body {
      margin: 0;
      padding: 0;
      font-family: 'Georgia', serif;
    }
    
    .certificate {
      width: 297mm;
      height: 210mm;
      padding: 40px;
      background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
      position: relative;
      box-sizing: border-box;
    }
    
    .certificate-inner {
      width: 100%;
      height: 100%;
      background: white;
      border: 20px solid #f7f7f7;
      box-shadow: 0 0 0 10px #667eea;
      padding: 60px;
      box-sizing: border-box;
      position: relative;
    }
    
    .certificate-header {
      text-align: center;
      margin-bottom: 40px;
    }
    
    .logo {
      font-size: 48px;
      font-weight: bold;
      color: #667eea;
      margin-bottom: 10px;
    }
    
    .certificate-title {
      font-size: 56px;
      font-weight: bold;
      color: #333;
      margin: 20px 0;
      text-transform: uppercase;
      letter-spacing: 4px;
    }
    
    .certificate-subtitle {
      font-size: 24px;
      color: #666;
      margin-bottom: 40px;
    }
    
    .certificate-body {
      text-align: center;
      margin: 40px 0;
    }
    
    .recipient-name {
      font-size: 48px;
      font-weight: bold;
      color: #667eea;
      margin: 30px 0;
      border-bottom: 3px solid #667eea;
      display: inline-block;
      padding-bottom: 10px;
    }
    
    .certificate-text {
      font-size: 22px;
      color: #555;
      line-height: 1.8;
      margin: 30px 0;
    }
    
    .highlight {
      color: #667eea;
      font-weight: bold;
    }
    
    .certificate-footer {
      display: flex;
      justify-content: space-between;
      margin-top: 60px;
      padding-top: 30px;
      border-top: 2px solid #eee;
    }
    
    .signature-block {
      text-align: center;
      flex: 1;
    }
    
    .signature-line {
      width: 200px;
      border-top: 2px solid #333;
      margin: 0 auto 10px;
    }
    
    .signature-name {
      font-size: 18px;
      font-weight: bold;
      color: #333;
    }
    
    .signature-title {
      font-size: 14px;
      color: #666;
    }
    
    .certificate-id {
      position: absolute;
      bottom: 20px;
      right: 40px;
      font-size: 14px;
      color: #999;
      font-family: 'Courier New', monospace;
    }
    
    .decorative-corner {
      position: absolute;
      width: 100px;
      height: 100px;
      border: 3px solid #667eea;
    }
    
    .corner-tl {
      top: 20px;
      left: 20px;
      border-right: none;
      border-bottom: none;
    }
    
    .corner-tr {
      top: 20px;
      right: 20px;
      border-left: none;
      border-bottom: none;
    }
    
    .corner-bl {
      bottom: 20px;
      left: 20px;
      border-right: none;
      border-top: none;
    }
    
    .corner-br {
      bottom: 20px;
      right: 20px;
      border-left: none;
      border-top: none;
    }
  </style>
</head>
<body>
  <div class="certificate">
    <div class="certificate-inner">
      <div class="decorative-corner corner-tl"></div>
      <div class="decorative-corner corner-tr"></div>
      <div class="decorative-corner corner-bl"></div>
      <div class="decorative-corner corner-br"></div>
      
      <div class="certificate-header">
        <div class="logo">iDigitalStudies</div>
        <div class="certificate-title">Certificate of Achievement</div>
        <div class="certificate-subtitle">This is to certify that</div>
      </div>
      
      <div class="certificate-body">
        <div class="recipient-name">${data.studentName}</div>
        
        <div class="certificate-text">
          has successfully completed the Digital Marketing program<br/>
          and has been placed at<br/>
          <span class="highlight">${data.companyName}</span><br/>
          as a<br/>
          <span class="highlight">${data.role}</span>
        </div>
        
        <div class="certificate-text" style="margin-top: 40px;">
          We congratulate ${data.studentName.split(' ')[0]} on this remarkable achievement<br/>
          and wish them continued success in their professional journey.
        </div>
      </div>
      
      <div class="certificate-footer">
        <div class="signature-block">
          <div class="signature-line"></div>
          <div class="signature-name">Director</div>
          <div class="signature-title">iDigitalStudies</div>
        </div>
        
        <div class="signature-block">
          <div style="margin-bottom: 20px;">
            <strong>Date:</strong> ${currentDate}
          </div>
          <div>
            <strong>Student ID:</strong> ${data.studentId}
          </div>
        </div>
        
        <div class="signature-block">
          <div class="signature-line"></div>
          <div class="signature-name">Head of Training</div>
          <div class="signature-title">iDigitalStudies</div>
        </div>
      </div>
      
      <div class="certificate-id">
        Certificate ID: ${data.certificateId}
      </div>
    </div>
  </div>
</body>
</html>
  `;
};

export const downloadCertificateAsHTML = (data: CertificateData) => {
  const html = generateCertificateHTML(data);
  const blob = new Blob([html], { type: 'text/html' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = `certificate_${data.certificateId}.html`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
};

export const openCertificateInNewTab = (data: CertificateData) => {
  const html = generateCertificateHTML(data);
  const blob = new Blob([html], { type: 'text/html' });
  const url = URL.createObjectURL(blob);
  window.open(url, '_blank');
  
  // Clean up after a delay
  setTimeout(() => {
    URL.revokeObjectURL(url);
  }, 1000);
};
